src/stats/pairhmm/homopolypairhmm.rs

```diff
@@ -0,0 +1,1145 @@
+use crate::stats::probs::LogProb;
+use crate::stats::Prob;
+
+/// The HomopolyPairHMM defined in this module has one Match state for each character from [A, C, G, T],
+/// for each of those Match states two corresponding Hop (homopolymer run) states
+/// (one for a run in sequence `x`, one for a run in `y`),
+/// as well as the usual GapX and GapY states.
+#[derive(Eq, PartialEq, Debug, Enum, Clone, Copy)]
+#[repr(usize)]
+pub enum State {
+    MatchA = 0,
+    MatchC = 1,
+    MatchG = 2,
+    MatchT = 3,
+    GapX = 4,
+    GapY = 5,
+    HopAX = 6,
+    HopAY = 7,
+    HopCX = 8,
+    HopCY = 9,
+    HopGX = 10,
+    HopGY = 11,
+    HopTX = 12,
+    HopTY = 13,
+}
+
+const STATES: [State; 14] = [
+    MatchA, MatchC, MatchG, MatchT, GapX, GapY, HopAX, HopAY, HopCX, HopCY, HopGX, HopGY, HopTX,
+    HopTY,
+];
+
+const MATCH_STATES: [State; 4] = [MatchA, MatchC, MatchG, MatchT];
+const HOP_X_STATES: [State; 4] = [HopAX, HopCX, HopGX, HopTX];
+const HOP_Y_STATES: [State; 4] = [HopAY, HopCY, HopGY, HopTY];
+
+// We define Shr (>>) for `State` such that a transition from State `a` to State `b` can be modeled
+// as `a >> b`, where `a >> b` is an integer in `0..(1 << (2 * NUM_STATES)) - 1]` used for indexing
+// the transition table (see `build_transition_table`).
+impl Shr for State {
+    type Output = usize;
+
+    fn shr(self, rhs: State) -> Self::Output {
+        let a = self as u32;
+        let b = rhs as u32;
+        interleave_bits(a, b) as usize
+    }
+}
+
+fn space_bits(a: u32) -> u64 {
+    let mut x = a as u64 & 0x0000_0000_FFFF_FFFF;
+    x = (x | (x << 16)) & 0x0000_FFFF_0000_FFFF;
+    x = (x | (x << 8)) & 0x00FF_00FF_00FF_00FF;
+    x = (x | (x << 4)) & 0x0F0F_0F0F_0F0F_0F0F;
+    x = (x | (x << 2)) & 0x3333_3333_3333_3333;
+    x = (x | (x << 1)) & 0x5555_5555_5555_5555;
+    x
+}
+
+fn interleave_bits(a: u32, b: u32) -> u64 {
+    space_bits(a) << 1 | space_bits(b)
+}
+
+/// Trait for parametrization of `PairHMM` hop behavior.
+pub trait HopParameters {
+    /// Probability to start hop in x.
+    fn prob_hop_x(&self) -> LogProb;
+
+    /// Probability to start hop in y.
+    fn prob_hop_y(&self) -> LogProb;
+
+    /// Probability to extend hop in x.
+    fn prob_hop_x_extend(&self) -> LogProb;
+
+    /// Probability to extend hop in y.
+    fn prob_hop_y_extend(&self) -> LogProb;
+}
+
+/// Trait for parametrization of `HomopolyPairHMM` emission behavior.
+pub trait EmissionParameters {
+    /// Emission probability for (x[i], y[j]).
+    /// Returns one of the enum variants Match(prob) or Mismatch(prob)
+    fn prob_emit_x_and_y(&self, s: State, i: usize, j: usize) -> XYEmission;
+
+    /// Emission probability for (x[i], -).
+    fn prob_emit_x_and_gap(&self, s: State, i: usize) -> LogProb;
+
+    /// Emission probability for (-, y[j]).
+    fn prob_emit_gap_and_y(&self, s: State, j: usize) -> LogProb;
+
+    /// Emission probability for (x[i], +).
+    fn prob_emit_x_and_hop(&self, s: State, i: usize) -> LogProb;
```

Comment:
Mhm, what does this mean? x is emitted while extending a hop in y, right? Shouldn't that be calculated from the HOPParameters only?

Reply:
Perhaps, … but then you wouldn't be able to adjust probabilities based on e.g. base qualities.

Reply:
Yes, indeed. This would also obviate the need of special EmissionParams wouldn't it?

Reply:
Seem tempting, but then we'd also lose the ability to define per-state emission probabilities (not strictly needed, but makes it easier to define for example different substitution probabilities for `C → G` etc).
Also, e.g. observing `AA` in state `MatchC` is highly unlikely, but quite likely in a general `Match` state. Without state dependent emission probabilities, we'd have to hardwire checks for cases like this.

Reply:
Observing AA in MatchC is impossible, isn't it? I have no problem with hardwiring that. Another option is having two levels of traits here. Keep the extended one internally, but hide it from the user, allowing him to specify the normal Emission trait.
I see these traits as meant for the user of the lib, and the user should not need to know about states at all.

Reply:
But the way the EmissionParameters trait is defined, It is impossible to check for specific bases  (since only indices `i` and `j` are supplied, the underlying sequence is hidden from the library's view)

Reply:
Yes, you are right! I will have to think about that...

Reply:
There's no convenient way to hide the `State` enum from users of the library.
I mean, we could have something like
```rust
trait EmissionParameters {
    fn emit_x_and_y(s: State, i : usize, j: usize) -> XYEmission {
        match s {
            MatchA => emit_x_and_y_in_A(i, j), 
            MatchC => emit_x_and_y_in_C(i, j), 
            MatchG => emit_x_and_y_in_G(i, j), 
            MatchT => emit_x_and_y_in_T(i, j), 
        }
    }
    fn emit_x_and_y_in_A(i: usize, j: usize) -> XYEmission;
    fn emit_x_and_y_in_C(i: usize, j: usize) -> XYEmission;
    fn emit_x_and_y_in_G(i: usize, j: usize) -> XYEmission;
    fn emit_x_and_y_in_T(i: usize, j: usize) -> XYEmission;
}
```
but that doesn't make it any less inconvenient

Reply:
Ok, now I know. You introduce a new trait
```
pub trait Emission {
    fn emit_x(i: usize) -> u8;
    fn emit_y(j: usize) -> u8;
}
```
and require the emission parameter arguments implements both `Emission` and `EmissionParameters`. Then, in your code, you use it to check what would be emitted for i and j, and only if the combination is valid you query the emission probability from `EmissionParameters`.

Reply:
Good idea!